jsonschema/validators.py

```diff
@@ -56,7 +56,7 @@ def _validates(cls):
     return _validates
 
 
-def create(meta_schema, validators=(), version=None, default_types=None):
+def create(meta_schema, validators=(), version=None, default_types=None, scope_key="id"):
```

Comment:
So, this makes me a bit uneasy (which is why I didn't do it immediately myself).

I think to be safe (future-proof) we need to make the interface a bit wider -- i.e., rather than just specifying the name of a key to use for scoping, we should provide something like a function here that the validator will use to detect scope (and yeah to make the current implementations just be lookups). I'm worried about the potential for some more complicated process, and once we already need an argument to do it, might as well have it be a generic one. Lemme know if that makes sense.

We also need to have direct tests for this (i.e. separate from any specific draft support).

Reply:
> Lemme know if that makes sense.
You mean specifying a validator for `id` or `$id` to handle the scope change?

I did consider that, however I wasn't sure the corresponding architectural change would be nice and wondered if a simpler solution is better at least until such time as there are future differences. But I see your point.

> We also need to have direct tests for this (i.e. separate from any specific draft support).
Agreed on that.

Reply:
I think it would be slightly narrower than a validator -- basically it is a function to replace the calls just to `someschema.get("id")` -- so, `scope_of(schema)`, say.

Reply:
And really, yeah this is tightly related to #346 unfortunately, which is to say, I think the even better thing is to come up with the right interface for *that*, provide the *RefResolver* class you want to use to this function, and have a function on that interface that is "scope of". I just can't decide whether it's obvious enough how that looks quite yet, and how much work it is to change it :/

Reply:
> provide the RefResolver class you want to use to this function

That was kinda where my initial thought was with this too. Unless I'm misunderstood, my concern is that a user may specify a `RefResolver` instance when calling `.validate()` -- this would mean a user is then specifying the draft number twice, i.e. `Draft4Validator` and `Draft4RefResovler`, which seemed unpleasant? I'm not up to speed enough to determine if that's a narrow enough issue though.

Reply:
Right so I think the idea there would be to stash what is passed in here as the resolver callable onto the validator interface -- i.e., the newer way to construct resolvers would become `DraftSomethingValidator(resolver=DraftSomethingValidator.Resolver(...))`

Reply:
Alright well I think I've got a bit of a better picture of the direction you'd like this to go in. I'll try and take a look at the related issues and then see if I can work towards something. Of course I shan't be offended if you're actively working on these things and prefer effort elsewhere.

Reply:
Ideas I have, time, much less so :(, especially given that the ideas are often only partially baked and need refinement or experimentation to see whether they fully fit -- so definitely definitely would appreciate the help!

I think the one thing that seems most clear at the moment is the stuff I mentioned about `type`, so I might recommend that one as the first self-contained area to poke at -- widen the type interface, get that merged, and then come back here and wiggle this onto the new one.

But there are many many paths forward and yeah I super appreciate any help.

Reply:
I've taken a look at the `type` stuff [1] - it's WIP and not yet ready for a PR, but if you have the time to give some feedback before I take it further it'd be really appreciated.

In particular:
- Rather than the single `is_type` injected function discussed in #365, I've taken the approach of a function per type. I figure this is more composable for extensions
- I presume a need to maintain backwards compatibility with the old `default_types` and per-instance `types` properties. I believe this does so (tests using this pass)  - but ideas appreciated
- I feel this works well for 'simple' types, for example handling the change in integer in draft6 and e.g.  a hypothetical scenario where strings encoding ints are valid under `integer` [2]. However, some difficulty remains for something like #364 to extend `object` to allow `namedtuple` - a user would have to provide custom validators for all `object`-based checks, e.g. in my test case [3]. Wether or not this is an acceptable level of work for a custom type or we should look to facilitate this I'm not sure. 

[1] https://github.com/Julian/jsonschema/compare/draft6...bsmithers:types
[2] https://github.com/bsmithers/jsonschema/blob/types/jsonschema/tests/test_types.py#L56
[3] https://github.com/bsmithers/jsonschema/blob/types/jsonschema/tests/test_types.py#L95

Reply:
So! I think that looks like a great start -- if I can inject one idea there (which is only 80% baked I think, so lemme know if I go off a rail) -- I think the right thing here is to take a cue from what's happening on the ref resolver and format checker sides -- to introduce a `TypeChecker` class, one that you associate an instance of with a validator class in create and extend.

The one major mistake I'd love to correct would be to not have mutable APIs for that -- so, to talk about the specific `namedtuple` case, the API I'd expect would look roughly like:

```
def object_or_namedtuple(value):
    return Draft4Validator.type_checker.is_type("object") or getattr(value, "_replace", None) is not None
```

say, and then `Draft4Validator.type_checker.redefine(object=object_or_namedtuple)` which should return a new instance (which you'd use via `Draft4Validator(type_checker=that)` where the appropriate redefinition has happened (`attr.evolve` + a `pyrsistent.pmap` of checkers are the pieces we'd use to make that happen).

You can then also have `type_checker.redefine(**types)` to convert from the old format, and I think the above does seem like a reasonable amount of work to do the namedtuple case.

How does that sound to you (and as usual apologies for spitting information out slowly here after you've already gotten started :), as usual with these things what seems right only starts to come out and evolve after progress has begun).